### PR TITLE
URLify the web UI

### PR DESCRIPTION
**Pull request by magopian**
Addresses #27 

- [x] URLify the "docs" and "builds" header buttons
- [x] URLify the filters
- [x] URLify specific builds

### CODE REVIEW

**Comment by mostlygeek** [comment]
Currently the URLs use the uuid. I think it's worthwhile to make the url something more predictable: 

Perhaps modeling after the AUS (balrog/updater service) urls is a good start: 

`https://aus5.mozilla.org/update/6/%PRODUCT%/%VERSION%/%BUILD_ID%/%BUILD_TARGET%/%LOCALE%/%CHANNEL%/%OS_VERSION%/%SYSTEM_CAPABILITIES%/%DISTRIBUTION%/%DISTRIBUTION_VERSION%/update.xml`

**Comment by Natim** [comment]
> Currently the URLs use the uuid.

It won't be the case anymore now that #30 landed.

**Comment by magopian** [comment]
Regarding https://github.com/mozilla-services/buildhub/pull/40#issuecomment-302804086 do we want to keep the same url structure? And thus the same "catalog categories"?

The question is: for the "catalog-ifying" of the UI (selecting a few things before displaying a huge list of builds, with a huge list of filters), what do we want to pre-select, and in which order?

Eg:
1/ select the product
2/ select the version
3/ select the platform

Or
1/ select the product
2/ select the channel
3/ select the version

Or a mix of those? Also, do we allow/want to select "all" for any of those?

**Comment by Natim** [comment]
What about a selection of the product and then a selection of which kind of channels, versions or platforms you want to filter on.

A bit like searching for bugs in Bugzilla

example2

**Comment by magopian** [comment]
I believe this is now ready for review: changing the url will update the filters, changing the filters will update the url. I've created two new issues to follow up on that:
- #64 Remove the "trees" filter which is redundant with the "channels" filter
- #65 Only use the `Route`, and drop the redundant `*Filter`s from the model

File: ui/Main.elm [hunk 1]
```diff
@@ -75,93 +69,11 @@ update msg ({ filterValues } as model) =
                         NewBuildIdSearch value ->
                             { model | buildIdFilter = value }
 
-                updatedFilteredBuilds =
-                    applyFilters updatedModelWithFilters
+                updatedRoute =
```
Comment (n1k0) [inline ui/Main.elm]: Nit: this is a `Model` value, `updatedRoute` sounds strange. `updatedModel` maybe?

File: ui/Url.elm [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+                    , channelFilter = "all"
+                    , platformFilter = "all"
+                    , versionFilter = "all"
+                    , localeFilter = "all"
+                }
+
+
+urlFromRoute : Model -> String
+urlFromRoute model =
+    case model.route of
```
Comment (n1k0) [inline ui/Url.elm]: Nit: If we're only using the `route` field, it sounds faire to just accept a `Route` argument, not a full `Model` one.

File: ui/Url.elm [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+                ++ locale
+
+        _ ->
+            "#/builds/"
+
+
+routeFromFilters : Model -> Route
+routeFromFilters model =
+    if model.buildIdFilter /= "" then
+        BuildIdView model.buildIdFilter
```
Comment (n1k0) [inline ui/Url.elm]: Nit: How about destructuring to decrease verbosity here?

```elm
routeFromFilters model =
    let
        { buildIdFilter, productFilter, channelFilter, platformFilter, versionFilter, localeFilter }
            = model
    in
        if buildIdFilter /= "" then
            BuildIdView buildIdFilter
        else if localeFilter /= "all" then
            LocaleView productFilter channelFilter platformFilter versionFilter localeFilter
```

Also it really feels like filters should be a `Dict`, but let's not waste time with this yet.

Comment (magopian) [inline ui/Url.elm]: I filed #65 already to remove them ;)

**Comment by magopian** [comment]
@n1k0 I believe I've addressed all your comments, thanks for the review! If you're happy with the changes, please let me know and I'll merge them.

**Comment by n1k0** [comment]
Review has been approved already, ship it! :)

Le 30 mai 2017 2:57 PM, "Mathieu Agopian" <notifications@github.com> a
écrit :

**Pull request merged**